### PR TITLE
Always add StateCheckpoint after execution

### PR DESCRIPTION
**Pull request by igor-aptos**
Previously, we were dynamically adding StateCheckpoint at the end of execution, if the block limit was reached.

We can clean up the code, and make it such that StateCheckpoint is not added to the input transactions, and is added after execution finishes (unconditionally). Since leaves the outside-visible behavior the same, it doesn't need a feature flag.

This is also needed for adding more information into StateCheckpoint itself.

### CODE REVIEW

**Comment by igor-aptos** [comment]
Current dependencies on/for this PR:
* main
  * **PR #10879** <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/10879?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
    * **PR #11110** <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/11110?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
      * **PR #10943** <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/10943?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
        * **PR #11287** <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/11287?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈
        * **PR #11253** <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/11253?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
        * **PR #11167** <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/11167?utm_source=stack-comment-icon" target="_blank">Graphite</a> 

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/aptos-labs/aptos-core/11287?utm_source=stack-comment).

File: execution/executor-types/src/lib.rs [hunk 1]
```diff
@@ -411,6 +411,42 @@ impl StateComputeResult {
...
+            .collect::<Vec<_>>();
+
+        assert!(
+            self.has_reconfiguration()
+                || matches!(output.last(), Some(Transaction::StateCheckpoint(_))),
+            "{:?}",
+            output.last()
+        );
+
+        println!(
```
Comment (danielxiangzl) [inline execution/executor-types/src/lib.rs:442]: Change log level?

File: consensus/consensus-types/src/executed_block.rs [hunk 2]
```diff
@@ -162,46 +162,19 @@ impl ExecutedBlock {
...
-                .compute_status()
-                .iter()
-                .position(|s| s.is_retry())
-            {
-                txns_with_state_checkpoint.insert(pos, Transaction::StateCheckpoint(self.id()));
-            } else {
-                txns_with_state_checkpoint.push(Transaction::StateCheckpoint(self.id()));
-            }
-        }
+        let input_txns = self
```
Comment (danielxiangzl) [inline consensus/consensus-types/src/executed_block.rs:172]: Add some comment for future readers? like `input_txns` contains blockmetadata, and `transactions_to_commit` will add state checkpoint.

File: execution/executor/src/components/apply_chunk_output.rs [hunk 3]
```diff
@@ -235,14 +229,7 @@ impl ApplyChunkOutput {
...
 
-        let status = if state_checkpoint_to_add.is_some() {
-            keeps_and_discards
-                .chain(once(TransactionStatus::Keep(ExecutionStatus::Success)))
-                .chain(retries)
-                .collect()
-        } else {
-            keeps_and_discards.chain(retries).collect()
-        };
+        let status = keeps_and_discards.chain(retries).collect();
```
Comment (danielxiangzl) [inline execution/executor/src/components/apply_chunk_output.rs:232]: Why removing this? State checkpoint is still inserted here but not in status?

File: execution/executor-types/src/lib.rs (hunk 1 above)
Comment (igor-aptos) [reply execution/executor-types/src/lib.rs:442]: I was just debugging, let me remove this.

File: execution/executor/src/components/apply_chunk_output.rs (hunk 3 above)
Comment (igor-aptos) [reply execution/executor/src/components/apply_chunk_output.rs:232]: as far as I can tell - status corresponds to input transactions (and why order of keeps and discards is kept)

for execution/commit, we use to_keep - which contains status itself.


this was the confusing part, we depended on status here being in the wrong place.

I've changed consensus/consensus-types/src/executed_block.rs, to append both the status and the new transaction at the same time, for consistency

Comment (danielxiangzl) [reply execution/executor/src/components/apply_chunk_output.rs:232]: But the state checkpoint txn is still added below in line 241?

Comment (igor-aptos) [reply execution/executor/src/components/apply_chunk_output.rs:232]: yes, that's what I mean. to_keep here seems to me to be the list of transactions that actually get committed

statuses here is always looked from the perspective of input transactions, and is kept in that order (not in the order they will be committed). let me rename statuses to statuses_for_input_txns?

Comment (danielxiangzl) [reply execution/executor/src/components/apply_chunk_output.rs:232]: Sounds good to me as long as the storage bosses are happy cc @msmouse @grao1991

File: types/src/transaction/mod.rs [hunk 4]
```diff
@@ -1233,6 +1233,38 @@ impl TransactionOutputProvider for TransactionOutput {
     }
 }
 
+#[derive(Debug)]
+pub struct BlockOutput<Output: Debug> {
```
Comment (grao1991) [inline types/src/transaction/mod.rs:1237]: Given this file is already too large, shall we put this struct into a separate file?

File: execution/executor/src/block_executor.rs [hunk 5]
```diff
@@ -383,6 +380,7 @@ where
             .sum();
         let target_version = ledger_info_with_sigs.ledger_info().version();
         if first_version + to_commit as u64 != target_version + 1 {
+            panic!();
```
Comment (grao1991) [inline execution/executor/src/block_executor.rs:383]: ?

File: types/src/transaction/mod.rs [hunk 6]
```diff
@@ -1836,6 +1868,16 @@ impl Transaction {
             Transaction::ValidatorTransaction(_) => String::from("validator_transaction"),
         }
     }
+
+    pub fn type_name(&self) -> &'static str {
```
Comment (grao1991) [inline types/src/transaction/mod.rs:1872]: where do you use this method?

File: execution/executor/src/components/apply_chunk_output.rs (hunk 3 above)
Comment (grao1991) [reply execution/executor/src/components/apply_chunk_output.rs:236]: How epoch ending is handled?

File: types/src/transaction/mod.rs (hunk 6 above)
Comment (igor-aptos) [reply types/src/transaction/mod.rs:1872]: in general, whenever I need to debug something, it is much easier to print the types of transaction, than everything. so nowhere committed, but don't want to be needing to be writing it over and over - it seems like a good utility to have committed.

File: execution/executor/src/components/apply_chunk_output.rs (hunk 3 above)
Comment (igor-aptos) [reply execution/executor/src/components/apply_chunk_output.rs:236]: epoch ending (i.e. reconfiguration) doesn't add StateCheckpoint/BlockPrologue transaction (i.e. adding transaction is gated upon that), i.e.:

let state_checkpoint_to_add =
new_epoch_marker.map_or_else(|| append_state_checkpoint_to_block, |_| None);

**Review by danielxiangzl — APPROVED** [review]

**Comment by igor-aptos** [comment]
ping for a second review :)

**Review by grao1991 — APPROVED** [review]

**Pull request merged**